### PR TITLE
add ignore_network_errors_at_startup

### PR DESCRIPTION
**Pull request by tagomoris**
This change is revival of #735, for v0.14 `out_forward`.
This parameter is used for the case: deployment systems will create the DNS records of hosts in `<server>`.

### CODE REVIEW

**Comment by tagomoris** [comment]
@repeatedly Could you review this change?

**Comment by tagomoris** [comment]
@repeatedly ping

**Comment by repeatedly** [comment]
Really?
I got following error and stopped timer.

```
2017-01-05 16:25:30 +0900 [info]: adding match pattern="forward.**" type="forward"
2017-01-05 16:25:30 +0900 [info]: adding forwarding server 'fooooobarrrr:24444' host="fooooobarrrr" port=24444 weight=60 plugin_id="object:3ff48a97a9dc"
2017-01-05 16:25:30 +0900 [warn]: failed to resolve node name when configured server="fooooobarrrr" error_class=SocketError error="getaddrinfo: nodename nor servname provided, or not known"
2017-01-05 16:25:30 +0900 [info]: adding match pattern="**" type="null"
2017-01-05 16:25:30 +0900 [info]: adding source type="forward"
2017-01-05 16:25:30 +0900 [info]: using configuration file: <ROOT>
  <source>
    @type forward
  </source>
  <match forward.**>
    @type forward
    flush_interval 5s
    ignore_network_errors_at_startup true
    retry_limit 4
    max_retry_wait 4
    <server>
      host "fooooobarrrr"
      port 24444
    </server>
    <buffer tag>
      flush_interval 5s
      retry_max_times 4
      retry_max_interval 4
    </buffer>
  </match>
  <match **>
    @type null
  </match>
</ROOT>
2017-01-05 16:25:30 +0900 [info]: starting fluentd worker pid=31724 ppid=31696 worker="0"
2017-01-05 16:25:30 +0900 [info]: delayed_commit_timeout is overwritten by ack_response_timeout
2017-01-05 16:25:30 +0900 [info]: fluentd worker is now running
2017-01-05 16:25:31 +0900 [error]: Unexpected error raised. Stopping the timer. title=:out_forward_heartbeat_request error_class=SocketError error="getaddrinfo: nodename nor servname provided, or not known"
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin/out_forward.rb:607:in `getaddrinfo'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin/out_forward.rb:607:in `resolve_dns!'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin/out_forward.rb:593:in `resolved_host'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin/out_forward.rb:569:in `send_heartbeat'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin/out_forward.rb:340:in `block in on_timer'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin/out_forward.rb:333:in `each'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin/out_forward.rb:333:in `on_timer'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin_helper/timer.rb:77:in `on_timer'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/cool.io-1.4.5/lib/cool.io/loop.rb:88:in `run_once'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/cool.io-1.4.5/lib/cool.io/loop.rb:88:in `run'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin_helper/event_loop.rb:77:in `block in start'
  2017-01-05 16:25:31 +0900 [error]: /Users/repeatedly/dev/fluentd/fluentd/lib/fluent/plugin_helper/thread.rb:66:in `block in thread_create'
2017-01-05 16:25:31 +0900 [error]: Timer detached. title=:out_forward_heartbeat_request
```

**Comment by tagomoris** [comment]
Oops, I misunderstood the problem... (and the former pull-request also doesn't solve this problem).

**Comment by tagomoris** [comment]
@repeatedly I pushed a commit for that case.

**Comment by tagomoris** [comment]
@repeatedly hey

File: lib/fluent/plugin/out_forward.rb [hunk 1]
```diff
@@ -550,9 +565,20 @@ def send_data(tag, chunk)
 
       # FORWARD_TCP_HEARTBEAT_DATA = FORWARD_HEADER + ''.to_msgpack + [].to_msgpack
       def send_heartbeat
+        begin
+          dest_addr = resolved_host
+          @resolved_once = true
+        rescue ::SocketError
+          if !@resolved_once && @ignore_network_errors_at_startup
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: `@ignore_network_errors_at_startup` is not a Node's instance variable.

File: lib/fluent/plugin/out_forward.rb [hunk 2]
```diff
@@ -550,9 +565,20 @@ def send_data(tag, chunk)
...
+            log.warn "failed to resolve node name in heartbeating", server: (n.name || n.host), error: e
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: `log` and `n` are not defined.

**Comment by tagomoris** [comment]
@repeatedly pushed updated commits.

**Comment by repeatedly** [comment]
LGTM

**Pull request merged**